File: frontend/src/models/pokemon.ts
export interface ListPokemonResponse {
    nextPage: string;
    page: number;
    pageSize: number;
    params: ListPokemonParams;
    pokemon: PokemonSpecies[];
    prevPage: string;
    total: number;
}

export interface PokemonSpecies {
    id: number;
    has_gender_differences: boolean;
    hatch_counter: number;
    is_baby: boolean;
    is_legendary: boolean;
    is_mythical: boolean;
    name: string;
    varieties: Partial<PokemonVariety>[];
}

export interface PokemonVariety {
    id: number;
    name: string;
    primary_type: MoveType;
    primary_type_id: number;
    secondary_type: MoveType | null;
    secondary_type_id: number | null;
    pokemon_species_id: number;
    sprite_url: string;
    cry: string;
    weight: number;
    height: number;
    is_default: boolean;
}


export interface MoveTypeResp {
    data: MoveType[]
    page: number
    pageSize: number
    total: number
}

export interface MoveType {
    id: number
    name: string
    double_damage_to?: number[]
    half_damage_to?: number[]
    no_damage_to?: number[]
    double_damage_from?: number[]
    half_damage_from?: number[]
    img_url?: string
    no_damage_from?: number[]
}

export interface ListPokemonParams {
    page?: number;
    pageSize?: number;
    pokemonName?: string | null
    pokemonTypeId?: number | null
}

File: api/models/pokemon.go
package models

type ListPokemonResponse struct {
	Pokemon  []PokemonSpecies  `json:"pokemon"`
	Page     int               `json:"page"`
	PageSize int               `json:"pageSize"`
	Total    int               `json:"total"`
	Params   ListPokemonParams `json:"params"`
	NextPage string            `json:"nextPage,omitempty"`
	PrevPage string            `json:"prevPage,omitempty"`
}

type Pokemon struct {
	ID               uint            `gorm:"primary_key" json:"id,omitempty"`
	Name             string          `gorm:"unique" json:"name,omitempty"`
	PrimaryType      *MoveType       `gorm:"foreignKey:PrimaryTypeId" json:"primary_type,omitempty"`
	PrimaryTypeId    *uint           `json:"primary_type_id,omitempty"`
	SecondaryType    *MoveType       `gorm:"foreignKey:SecondaryTypeId;" json:"secondary_type,omitempty"`
	SecondaryTypeId  *uint           `json:"secondary_type_id,omitempty"`
	PokemonSpecies   *PokemonSpecies `gorm:"foreignKey:PokemonSpeciesId" json:"pokemon_species,omitempty"`
	PokemonSpeciesId uint            `json:"pokemon_species_id,omitempty"`
	SpriteUrl        string          `json:"sprite_url,omitempty"`
	Cry              string          `json:"cry,omitempty"`
	Weight           int             `json:"weight,omitempty"`
	Height           int             `json:"height,omitempty"`
	IsDefault        bool            `json:"is_default,omitempty"`
	LearnableMoves   []Move          `gorm:"many2many:pokemon_move_sets;" json:"learnable_moves,omitempty"`
}

//TODO: add species JSONb column to table

type Sprites struct {
	BackDefault      string `json:"back_default"`
	BackFemale       string `json:"back_female,omitempty"`
	BackShiny        string `json:"back_shiny"`
	BackShinyFemale  string `json:"back_shiny_female,omitempty"`
	FrontDefault     string `json:"front_default"`
	FrontFemale      string `json:"front_female,omitempty"`
	FrontShiny       string `json:"front_shiny"`
	FrontShinyFemale string `json:"front_shiny_female,omitempty"`
}

type PokemonSpecies struct {
	ID                   uint      `gorm:"primary_key" json:"id,omitempty"`
	HasGenderDifferences bool      `json:"has_gender_differences,omitempty"`
	HatchCounter         int       `json:"hatch_counter,omitempty"`
	IsBaby               bool      `json:"is_baby,omitempty"`
	IsLegendary          bool      `json:"is_legendary,omitempty"`
	IsMythical           bool      `json:"is_mythical,omitempty"`
	Name                 string    `json:"name,omitempty"`
	Varieties            []Pokemon `json:"varieties,omitempty"`
}

type ListPokemonParams struct {
	PokemonName   string `form:"pokemonName" json:"pokemonName,omitempty"`
	PokemonTypeId int64  `form:"pokemonTypeId" json:"pokemonTypeId,omitempty"`
}
